### PR TITLE
Add an exported counter to track configuration reloads

### PR DESCRIPTION
**Pull request by swsnider**
This makes it easier for automation to detect whether a configuration reload actually worked, so that rollbacks of config at the filesystem can be done in concert with rollbacks in memory.

### CODE REVIEW

File: cmd/prometheus/main.go [hunk 1]
```diff
@@ -174,11 +183,13 @@ type Reloadable interface {
 
 func reloadConfig(filename string, rls ...Reloadable) bool {
 	log.Infof("Loading configuration file %s", filename)
+	configAttempts.WithLabelValues("attempted").Add(1)
```
Comment (brian-brazil) [inline cmd/prometheus/main.go]: This isn't a partition over the space, this should be two metrics - attempts and successes.

The last unixtime a config reload succeeded would probably also be useful.

Comment (fabxc) [inline cmd/prometheus/main.go]: Yes, the common approach to those tried/failed/succeeded metrics is this:
https://github.com/prometheus/prometheus/blob/master/retrieval/discovery/dns.go#L42-L53

**Comment by swsnider** [comment]
PTAL -- I addressed these concerns and verify that it works as expected.

File: cmd/prometheus/main.go [hunk 2]
```diff
@@ -46,6 +46,24 @@ func main() {
...
+		Name:      "configuration_successes",
+		Help:      "number of times a configuration reload was successful.",
+	})
+	configSuccessTime = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: "prometheus",
+		Name:      "configuration_last_success",
+		Help:      "Timestamp of the last successful configuration reload.",
+	})
+)
+
```
Comment (juliusv) [inline cmd/prometheus/main.go]: Metric-naming- and doc-string-wise it'd be good to make this closer to our conventions at http://prometheus.io/docs/practices/naming/ (for example, counters should end with `_total`) and other code. I'd also include `reload` in the names to make it clearer what it's about. Something like this:

```
var (
	configAttempts = prometheus.NewCounter(prometheus.CounterOpts{
		Namespace: "prometheus",
		Name:      "config_reload_attempts_total",
		Help:      "Number of times a configuration reload was attempted.",
	})
	configSuccesses = prometheus.NewCounter(prometheus.CounterOpts{
		Namespace: "prometheus",
		Name:      "config_reload_successes_total",
		Help:      "Number of times a configuration reload was successful.",
	})
	configLastSuccessTime = prometheus.NewGauge(prometheus.GaugeOpts{
		Namespace: "prometheus",
		Name:      "config_last_reload_success_timestamp_seconds",
		Help:      "Unix timestamp in seconds of the last successful configuration reload.",
	})
)
```

When I think about what I'd want to monitor in terms of config reload failures, it's usually whether the last configuration reload was a successful one (which includes the one that always happens at startup). So maybe a boolean `1` / `0` gauge like `config_last_reload_successful` would make sense? Then you can alert if that gauge stays `0` for too long. @brian-brazil opinions?

Comment (brian-brazil) [inline cmd/prometheus/main.go]: > So maybe a boolean 1 / 0 gauge like config_last_reload_successful would make sense? Then you can alert if that gauge stays 0 for too long.

Yes, that's what I was generally thinking (and having that also on the status page as a big warning with the last error would be good).

Comment (juliusv) [inline cmd/prometheus/main.go]: :+1: for that, though the latter would be out of scope for this PR :)

Do we even want to keep the last success timestamp metric then? It feels like the `config_last_reload_successful` metric would be more useful than the last successful timestamp in the use cases I can think of right now.

Comment (brian-brazil) [inline cmd/prometheus/main.go]: Knowing how old your config is is useful.

Comment (juliusv) [inline cmd/prometheus/main.go]: Ok, if you view it from that angle, yeah :) Let's keep it then.

Comment (swsnider) [inline cmd/prometheus/main.go]: OK, in progress

**Comment by swsnider** [comment]
PTAL

File: cmd/prometheus/main.go [hunk 3]
```diff
@@ -46,6 +46,19 @@ func main() {
...
 
+var (
+	configSuccess = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: "prometheus",
+		Name:      "config_last_reload_successful",
+		Help:      "Whether the last configuration reload attempt was successful.",
+	})
+	configSuccessTime = prometheus.NewGauge(prometheus.GaugeOpts{
+		Namespace: "prometheus",
+		Name:      "configuration_last_reload_success_timestamp_seconds",
```
Comment (brian-brazil) [inline cmd/prometheus/main.go]: Be consistent on config vs. configuration

Comment (swsnider) [inline cmd/prometheus/main.go]: Done.

File: cmd/prometheus/main.go [hunk 4]
```diff
@@ -186,6 +202,12 @@ func reloadConfig(filename string, rls ...Reloadable) bool {
 	for _, rl := range rls {
 		success = success && rl.ApplyConfig(conf)
 	}
+	if success {
+		configSuccess.Set(1.0)
+		configSuccessTime.Set(float64(time.Now().Unix()))
+	} else {
+		configSuccess.Set(0)
```
Comment (brian-brazil) [inline cmd/prometheus/main.go]: this is redundant, probably best to combine this if with the one above

Comment (juliusv) [inline cmd/prometheus/main.go]: If you want to get rid of the multiple places where you have to remember to set the metric to the right value, the best would be to use a named return value:

```go
func reloadConfig(filename string, rls ...Reloadable) (success bool) {
```

...and then do this at the beginning of the function:

```go
defer func() {
  if success {
    configSuccess.Set(1)
    configSuccessTime.Set(float64(time.Now().Unix()))
  } else {
    configSuccess.Set(0)
  }
}()
```

Comment (juliusv) [inline cmd/prometheus/main.go]: (and then in line 200, `success := true` would have to be changed to `success = true` to not shadow the named return variable)

Comment (swsnider) [inline cmd/prometheus/main.go]: Done

**Comment by brian-brazil** [comment]
:+1:

**Comment by fabxc** [comment]
:+1: after squashing
Thanks a lot!

**Comment by swsnider** [comment]
Squashed!

**Pull request merged**